### ISSUE TITLE
Over CPU and RAM usage while downloading files from S3

### ISSUE THREAD

**Issue by Toanzzz** [nmcapule/metabox-go#4]
While downloading file from S3, the CPU & memory usage got extremely high, 100% and around 2GB+ for ~250MB file.
This is pretty weird consider it only downloading, not yet come to decompressing it yet.

So as this problem exists I couldn't make it run inside limited environment like small EC2 instance or inside docker container

**Comment by nmcapule** [comment]
HI! Can you provide more context? Afaik memory leaks are only relevant on long-running applications.

go-metabox is a CLI tool that exits as soon as it's done doing its job. Once the tool has exited, it is up to the Go runtime to cleanup memory. The OS itself also tracks memory allocations of a program, so if the program has exited -- the OS will also cleanup the dangling bits.

Are you encountering OOM errors?

**Comment by Toanzzz** [comment]
Sorry, I was in rush and didn't describe the problem properly. I updated the PR and description

**Comment by Toanzzz** [comment]
Btw I really glad that this tool haven't abandoned yet 😀
Willing to help regarding this if you want (maybe this weekend)

**Comment by nmcapule** [comment]
No prob! Can you post your *.metabox.yaml file for reference? Make sure that there are no sensitive values for `access_key_id` and `secret_access_key` though :)

**Comment by Toanzzz** [comment]
sure, here it is
```yaml
version: 0.1
workspace:
  user_identifier: "${TARGET_USER}"
  tags_generator: ["author:${TARGET_USER}"]
target:
  prefix_path: ./dumps
  includes: ["**/*"]
  excludes: [".git/"]
backups:
  - driver: s3
    s3:
      access_key_id: ${AWS_ACCESS_KEY_ID}
      secret_access_key: ${AWS_SECRET_ACCESS_KEY}
      region: ${AWS_DEFAULT_REGION}
      bucket: ${AWS_S3_BUCKET}
      prefix_path: ${AWS_S3_PREFIX}

```

**Comment by nmcapule** [comment]
Thank you. I've reproduced the bug on a 1G Linode instance, showing an OOM error:

```
fatal error: runtime: out of memory                                                                                                                                     
                                                                                                                                                                        
runtime stack:                                                                                                                                                          
runtime.throw(0xb9c21c, 0x16)                                                                                                                                           
        /home/linuxbrew/.linuxbrew/Cellar/go/1.16.5/libexec/src/runtime/panic.go:1117 +0x72                                                                             
runtime.sysMap(0xc04c000000, 0x10000000, 0x10b1530)                                                                                                                     
        /home/linuxbrew/.linuxbrew/Cellar/go/1.16.5/libexec/src/runtime/mem_linux.go:169 +0xc6                                                                          
runtime.(*mheap).sysAlloc(0x1098340, 0xcc00000, 0x42dbb7, 0x1098348)                                                                                                    
        /home/linuxbrew/.linuxbrew/Cellar/go/1.16.5/libexec/src/runtime/malloc.go:729 +0x1e5                                                                            
runtime.(*mheap).grow(0x1098340, 0x6403, 0x0)                                                                                                                           
        /home/linuxbrew/.linuxbrew/Cellar/go/1.16.5/libexec/src/runtime/mheap.go:1346 +0x85                                                                             runtime.(*mheap).allocSpan(0x1098340, 0x6403, 0x460100, 0x1098340)                                                                                                      
        /home/linuxbrew/.linuxbrew/Cellar/go/1.16.5/libexec/src/runtime/mheap.go:1173 +0x609                                                                            runtime.(*mheap).alloc.func1()                                                      
        /home/linuxbrew/.linuxbrew/Cellar/go/1.16.5/libexec/src/runtime/mheap.go:910 +0x59                                                                              runtime.systemstack(0x46b3b4)                                                       
        /home/linuxbrew/.linuxbrew/Cellar/go/1.16.5/libexec/src/runtime/asm_amd64.s:379 +0x66                                                                           runtime.mstart()          
        /home/linuxbrew/.linuxbrew/Cellar/go/1.16.5/libexec/src/runtime/proc.go:1246     
```

It happens because the S3 storage client implementation in metabox is naively loading the whole file into memory before putting it into a file.
https://github.com/nmcapule/metabox-go/blob/1b81629d8c9b2f47ee1cc68fd7b1e0bd2abd9850/storage/s3.go#L64

Created PR: https://github.com/nmcapule/metabox-go/pull/5

**Comment by nmcapule** [comment]
Merged #5
